pkg/dataobj/consumer/partition_processor_test.go

```diff
@@ -0,0 +1,310 @@
+	"io"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/require"
+	"github.com/thanos-io/objstore"
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	"github.com/grafana/loki/v3/pkg/dataobj"
+	"github.com/grafana/loki/v3/pkg/dataobj/uploader"
+	"github.com/grafana/loki/v3/pkg/logproto"
+
+	"github.com/grafana/loki/pkg/push"
+)
+
+// mockBucket implements objstore.Bucket interface for testing
+type mockBucket struct {
+	uploads map[string][]byte
+	mu      sync.Mutex
+}
+
+func newMockBucket() *mockBucket {
+	return &mockBucket{
+		uploads: make(map[string][]byte),
+	}
+}
+
+func (m *mockBucket) Close() error                             { return nil }
+func (m *mockBucket) Delete(_ context.Context, _ string) error { return nil }
+func (m *mockBucket) Exists(_ context.Context, name string) (bool, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	_, exists := m.uploads[name]
+	return exists, nil
+}
+func (m *mockBucket) Get(_ context.Context, name string) (io.ReadCloser, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	data, exists := m.uploads[name]
+	if !exists {
+		return nil, errors.New("object not found")
+	}
+	return io.NopCloser(bytes.NewReader(data)), nil
+}
+func (m *mockBucket) GetRange(_ context.Context, _ string, _, _ int64) (io.ReadCloser, error) {
+	return nil, nil
+}
+func (m *mockBucket) Upload(_ context.Context, name string, r io.Reader) error {
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.uploads[name] = data
+	return nil
+}
+func (m *mockBucket) Iter(_ context.Context, _ string, _ func(string) error, _ ...objstore.IterOption) error {
+	return nil
+}
+func (m *mockBucket) Name() string { return "mock" }
+func (m *mockBucket) Attributes(_ context.Context, _ string) (objstore.ObjectAttributes, error) {
+	return objstore.ObjectAttributes{}, nil
+}
+func (m *mockBucket) GetAndReplace(_ context.Context, name string, _ func(io.Reader) (io.Reader, error)) error {
+	return m.Upload(context.Background(), name, bytes.NewReader([]byte{}))
+}
+func (m *mockBucket) IsAccessDeniedErr(_ error) bool {
+	return false
+}
+func (m *mockBucket) IsObjNotFoundErr(err error) bool {
+	return err != nil && err.Error() == "object not found"
+}
+func (m *mockBucket) IterWithAttributes(_ context.Context, _ string, _ func(objstore.IterObjectAttributes) error, _ ...objstore.IterOption) error {
+	return nil
+}
+func (m *mockBucket) Provider() objstore.ObjProvider {
+	return objstore.ObjProvider("MOCK")
+}
+func (m *mockBucket) SupportedIterOptions() []objstore.IterOptionType {
+	return nil
+}
+
+var testBuilderConfig = dataobj.BuilderConfig{
+	TargetPageSize:    2048,
+	TargetObjectSize:  4096,
+	TargetSectionSize: 4096,
+
+	BufferSize: 2048 * 8,
+}
+
+// TestIdleFlush tests the idle flush behavior of the partition processor
+// under different timeout and initialization conditions.
+func TestIdleFlush(t *testing.T) {
```

Comment:
It's probably worth adding a `t.Parallel()` to these tests so they run concurrently, since they rely on the time. 
In the future, we might be able to find a way to inject the time so we don't need to rely on sleeping, but lets save that for another iteration.